source/dem/pw_contact_force.cc

```diff
@@ -98,5 +98,130 @@ PWContactForce<dim>::update_contact_information(
+      // Second's law of Newton says that the normal reaction is equal to all
+      // forces in the other direction
+      N[2] = -(Sum_F[2] + P[2]);
+      // std::cout << "\nN " << N << "\n";
+      bool is_moving = (abs(Sum_F[1] + P[1]) > abs(mu_s * N[2]));
+      // std::cout << "\nis_moving " << is_moving << "\n";
+      std::string rotation_direction;
+      if (is_moving)
+        {
+          rotation_direction =
+            (Sum_F[1] + P[1] > abs(mu_s * N[2])) ? "left" : "right";
+        }
+      if (is_moving && rotation_direction == "right")
+        {
+          F_friction[1] = -mu_k * N[2];
+        }
+      else if (is_moving && rotation_direction == "left")
+        {
+          F_friction[1] = mu_k * N[2];
+        }
+      else if (!is_moving)
+        {
+          F_friction[1] = -(P[1] + Sum_F[1]);
+        }
+      // std::cout << "\nF_friction " << F_friction << "\n";
+      Tensor<1, dim> vector_contact;
+      vector_contact[2] = -radius;
+      M_friction        = cross_product_3d(vector_contact, F_friction);
+      // std::cout << "\nM_friction " << M_friction << "\n";
+      Tensor<1, dim> add_F, add_M, transition;
+      add_F = (dt / m) * (Sum_F + F_friction + P + N);
+      // std::cout << "\nadd_F " << add_F << "\n";
+      Tensor<1, dim> vector_rolling_contact;
+      vector_rolling_contact[1] = mu_r * radius;
+      vector_rolling_contact[2] = -radius;
+      Tensor<1, dim> M_rolling(cross_product_3d(vector_rolling_contact, N));
+      for (int i = 0; i < dim; i++)
+        {
+          add_M[i] = (dt / I[i]) * (Sum_M[i] + M_friction[i] + M_rolling[i]);
+        }
+      // std::cout << "\nadd_M " << add_M << "\n";
+      for (const auto &it : boundary_translational_velocity_map)
+        {
+          boundary_translational_velocity_map[it.first] = it.second + add_F;
+          // std::cout << "\nboundary_translational_velocity_map[it.first] " <<
+          // boundary_translational_velocity_map[it.first] << "\n";
+          transition = boundary_rotational_speed_map[it.first] *
+                         boundary_rotational_vector[it.first] +
+                       add_M;
+          boundary_rotational_speed_map[it.first] = transition.norm();
+          // std::cout << "\nboundary_rotational_speed_map[it.first] " <<
+          // boundary_rotational_speed_map[it.first] << "\n";
+          boundary_rotational_vector[it.first] = transition / transition.norm();
+          // std::cout << "\nboundary_rotational_vector[it.first] " <<
+          // boundary_rotational_vector[it.first] << "\n";
+        }
+    }
+}
+
+template <int dim>
+Tensor<1, dim>
+PWContactForce<dim>::base_change(const Tensor<1, dim> tensor,
+                                 const double         theta)
+{
+  Tensor<1, dim> out(tensor);
+  out[1] =
+    tensor[1] * cos(theta * M_PI / 180) - tensor[2] * sin(theta * M_PI / 180);
+  out[2] =
+    tensor[2] * cos(theta * M_PI / 180) + tensor[1] * sin(theta * M_PI / 180);
+  return out;
+}
+
+
```

Comment:
What is this for?

Reply:
This is to change the base of every forces that is not in the simplified axis. We can talk about this